### ISSUE TITLE
Suggestion: Partial Filter Matches

### ISSUE THREAD

**Issue by dabros** [wenzhixin/bootstrap-table#678]
So im doing this as part of a demo webapp, no real time to create github version and pretty up all my other changes (there are several)

However, this improvement is simple, it enhances the behaviour for filterBy (#281) and allows partial matches and takes into account multiple columns/fields nicely.

This was adapted from the 1.6.0 release

**EDIT**: added line for options.undefinedText

#### bootstrap-table.js initSearch enhancement

 ```javascript

    BootstrapTable.prototype.initSearch = function () {
        var that = this;

        if (this.options.sidePagination !== 'server') {
            var s = this.searchText && this.searchText.toLowerCase();
            var f = $.isEmptyObject(this.filterColumns) ? null: this.filterColumns;

            /// console.log('bootstrapTable.initSearch', { s: s, f: f });

            // Check filter
            this.data = f ? $.grep(this.options.data, function (item, i) {

                var passCount = 0;
                for (var key in f) {
                    key = $.isNumeric(key) ? parseInt(key, 10) : key;
                    var value = item[key];

                    var fval = f[key].toLowerCase().trim();
                    value = calculateObjectValue(that.header,
                        that.header.formatters[$.inArray(key, that.header.fields)],
                        [value, item, i, key], value);


                    // If fval is EXACTLY the same as options.undefinedText value, then find empty values
                    if ((that.options.undefinedText === fval) && (value === null))
                        passCount++
                    // Else check for partial matches of each fval
                    else if ((typeof value === 'string' || typeof value === 'number') &&
                        (value + '').toLowerCase().indexOf(fval) !== -1 )
                        passCount++;


                    /// DEBUG ///
                    /// console.info('filter debug', { value: value, fval: fval, item: item, i: i, f: f, key: key, 'that.header': that.header, passCount: passCount });

                }

                /// NOTE - f.size solution from http://stackoverflow.com/a/5527037/3770004
                ///      - if issues with hasOwnProperty or other, use a second int++ inside loop
                return passCount >= Object.keys(f).length;
              
            }) : this.options.data;


            this.data = s ? $.grep(this.data, function (item, i) {
                for (var key in item) {
                    key = $.isNumeric(key) ? parseInt(key, 10) : key;
                    var value = item[key];

                    // Fix #142: search use formated data
                    value = calculateObjectValue(that.header,
                        that.header.formatters[$.inArray(key, that.header.fields)],
                        [value, item, i, key], value);

                    var index = $.inArray(key, that.header.fields);
                    if (index !== -1 && that.header.searchables[index] &&
                        (typeof value === 'string' ||
                        typeof value === 'number') &&
                        (value + '').toLowerCase().indexOf(s) !== -1) {
                        return true;
                    }
                }
                return false;
            }) : this.data;
        }
    };
```

#### UI Implementation

This is only for 'bonus credit', but works great for me and if merged could definitely add to what is currently an underpowered feature.

I looked at http://bootsnipp.com/snippets/featured/panel-table-with-filters-per-column for design inspiration, and used a similar approach to create a filtertoggle button and replace th.th-inner with input.disabled elements.

``` html

 <div id="@Model.ajax_itemName-custom-toolbar" style=''> 
        <button class="btn btn-default btn-sm pull-left" type="button" name="filtertoggle" title="Toggle Filters" style="margin-right:5px;"><i class="glyphicon glyphicon-filter icon-filter"></i></button>
 </div>
   ...
   <!-- CHANGED: added  'data-filters="true" ' to table, but atm only picked up by event callback code -->
```

Took some changes to css and some slight tweaks to allow a 'onPostInit' (run only ever once) event so i could setup this and other ui, but i pass the button using the 'data-toolbar' option and the js looks something like this:

 ```javascript

  /// NOTE - for this share i left in the several different ways to check for .btn-filter (such as name, data-attribute and classname)
  ///            - in actual implementation, checks against data-filters should enable, and $wrap.find() should be used as sanity check (and to remove ui if not needed)

            // NOTE IMPORTANT - custom event that only fires once ever, after all table UI constructed
            onPostInit: function ($_container, _header) {

                // ...
           
                if (!!$tt.data('filters') && $wrap.find('.btn-filter')) {
                    $tt.addClass('my-filterable-table');
                    $wrap.find('.btn[name=filtertoggle]').on('click', function () {
                        $filters = $tt.find('thead .my-filterable-input');
                        // $tbody = $tt.find('.table tbody');
                        if ($filters.prop('disabled') == true) {
                            $filters.prop('disabled', false).addClass('input-sm');
                            $filters.first().focus();
                        } else {
                            $filters.val('').prop('disabled', true).removeClass('input-sm');
                            //$tbody.find('.no-result').remove();
                            //$tbody.find('tr').show();
                        }
                    });
                }
               // ...
           },

            /// NOTE - header can be refreshed by toggleView and other calls, so need to update accordingly
            onPostHeader: function ($_header, _header) {
                if (!!$tt.data('filters'))
                {
                    /// console.log('onPostHeader', [$_header, _header]);
                    $_header.find('.th-inner').each(function (th_i) {
                        // Sanity check if needed
                        if (!$(this).find('input').length) {
                            var colPlaceholder = $(this).text();
                            var colName = _header.fields[th_i];
                            var $colInput = $('<input type="text" name="' + colName + '" class="form-control my-filterable-input XXinput-sm" placeholder="' + colPlaceholder + '" disabled/>')
                                .off('keyup change').on('keyup change', function () {
                                    var _filterdata = {}; // $tt.find('.my-filterable-table-tr .my-filterable-input').serializeArray();
                                    $_header.find('.my-filterable-input').each(function () {
                                        if ($(this).val().trim() !== "")
                                            _filterdata[$(this).attr('name')] = $(this).val().trim();
                                    });
                                    $tt.bootstrapTable('filterBy', _filterdata);
                                    console.log('my-filterable-input keyup|change', _filterdata);
                                }).on('click', function (e) { e.stopPropagation(); });
                            $(this).empty().append($colInput);
                        }
                    });
                }
            },
```

This code is part of my alpha mod of this, which includes many other tweaks, and implemented using the 'onPostBody' event (which was documented but not in 1.6.0, so i added it). 

Some of the changes to bootstrap-table.js are:

``` javascript

    BootstrapTable.DEFAULTS = {
        // ...
        /// CHANGED NEW - match new events to functions     
        , onPostToolbar: function () { return false; }
        , onPostPagination: function ($_container) { return false; }
        , onPostInit: function ($_container, _header) { return false; }
        , onPostHeader: function ($_header, _header) { return false; }
    };


    BootstrapTable.EVENTS = {
        // ...
        /// CHANGED NEW - match new events to functions
        , 'post-toolbar.bs.table': 'onPostToolbar'
        , 'post-pagination.bs.table': 'onPostPagination'
        , 'post-init.bs.table': 'onPostInit'
        /// NOTE - be careful of unsynced user-doc, as doco existed but no function
        , 'post-header.bs.table': 'onPostHeader'

    };

    BootstrapTable.prototype.init = function () {
        // ...
        /// CHANGED NEW - called only once ever, and passes container and header info out
        this.trigger('post-init',this.$container, this.header);
    };
   
   // ...
 
   BootstrapTable.prototype.onSort = function (event) {
    // ...
    /// CHANGED - change to prepend so doesnt interfere with my-filterable-table
     $this.add($this_).data('order', this.options.sortOrder)
        .find('.th-inner').prepend(this.getCaretHtml());
   // ... 
   };




    BootstrapTable.prototype.togglePagination = function () {
        this.options.pagination = !this.options.pagination;
        /// CHANGED - need more obvious 'active' state for this toggle button
        var button = this.$toolbar.find('button[name="paginationSwitch"]');
        var icon = button.find('i');
        if (this.options.pagination) {
            button.removeClass('active');
            icon.attr("class", this.options.iconsPrefix + " " + this.options.icons.paginationSwitchDown);
        } else {
            button.addClass('active');
            icon.attr("class", this.options.iconsPrefix + " " + this.options.icons.paginationSwitchUp);
        }
        this.updatePagination();
    };


```

I know from previous projects id never get time to make proper github branch, especially since id have to remove most my other tweaks, so thought this might work to share it.

If this does help, please say so, as I'm constantly extending base plugins and never getting time to properly share my changes - if this does help maybe its at least some way i can give back.

Loving this plugin so far - keep it up :+1:

**Comment by dabros** [comment]
Ideally we would have 'showColumnFilters' or similar option and data-attribute. 

Then that would allow option for this, as is I just ran this as an event callback and used some css tweaks, changed append 'span.order' to prepend 'span.order' and e.stopPropagation to not interfere with the sortable code.

**Comment by dabros** [comment]
As @djhvscf stated elsewhere, there is the new filterControl option (available through extension atm i believe) that can answer this

To see example of MY implementation, see my post in #725 and the jsfiddle i share - not sure where to direct you to for good example of filterControl

P.S. as i say above though, the UI implementation is mostly 'bonus credit' - the small change to filterBy itself will allow partial matches for custom UI (like mine) or any calls via other means.

**Comment by djhvscf** [comment]
@dabros  actually I like your post of filterControl I would like to do some modification on that extension.. Really it is so good! @dabros

**Comment by teyou** [comment]
It seems there is a problem when I try to search on multiple fields.
Even though one of the field's value has partially match, it doesn't show in the result.

Hence, my solution is 

within 'BootstrapTable.prototype.initSearch' I changed this line 
```
      return passCount >= Object.keys(f).length;
```
to
```
return passCount >0 && passCount <= Object.keys(f).length;
```

**Comment by dabros** [comment]
@teyou - doesnt sound like what i was posting about, so pls remove your comment and create new issue/topic or find more appropriate location
